qtrees/data_processor.py

```diff
@@ -0,0 +1,311 @@
+from typing import Union, Optional
+from sklearn.preprocessing import OrdinalEncoder
+
+ROLLING_WINDOW = 7
+FORECAST_HORIZON = 14
+AUTOREGRESSIVE_LAG = 3
+
+class Data_loader:
+    def __init__(self, engine):
+        self.engine = engine
+
+    def download_nowcast_inference_data(self, date: dt.date = dt.date.today() - dt.timedelta(days=1), public_run: bool = False, batch_size: Optional[int] = None, batch_num: Optional[int] = None):
+        # TODO: Figure out a way to download data in batches s.t. we do not have to hold that much info in memory
+        self.forecast = False
+        self.date = date
+        self.public_run = public_run
+        self.batch_size = batch_size
+        self.batch_num = batch_num
+        data = self._download_data(with_sensors=False)
+        if data is not None:
+            data = data.merge(self._get_weather_measurements(), how="left", left_on="timestamp", right_index=True)
+        return data
+    
+    def download_forecast_inference_data(self, date: dt.date = dt.date.today(), public_run: bool = False, batch_size: Optional[int] = None, batch_num: Optional[int] = None):
+        self.forecast = True
+        self.date = date
+        self.public_run = public_run
+        self.batch_size = batch_size
+        self.batch_num = batch_num
+        data = self._download_data(with_sensors=False)
+        return data
+    
+    def download_nowcast_training_data(self, public_run: bool = False):
+        self.forecast = False
+        self.public_run = public_run
+        self.batch_size = None
+        data = self._download_data(with_sensors=True)
+        data = data.merge(self._get_weather_measurements(), how="left", left_on="timestamp", right_index=True)
+        return data
+    
+    def download_forecast_training_data(self, public_run: bool = False):
+        self.forecast = True
+        self.public_run = public_run
+        self.batch_size = None
+        data = self._download_data(with_sensors=True)
+        data = data.merge(self._get_weather_measurements(), how="left", left_on="timestamp", right_index=True)
+        return data
+
+    def _download_data(self, with_sensors: bool = True) -> pd.DataFrame:
+        '''Download data for training or inference. If sensordata is downloaded, we ignore tree_subset and date.
+        Otherwise we can download data for all trees for one day for inference. For evaluation we need to split training data into train and eval, e.g. via Crossvalidation'''
+        self.with_sensors = with_sensors
+        if self.with_sensors:
+            self.date = None
+        # Select all trees
+        if self.batch_size is None:
+            trees = pd.read_sql("SELECT id,gattung,standalter FROM public.trees WHERE street_tree = true", self.engine.connect())
+        else:
+            trees = pd.read_sql(f"SELECT id,gattung,standalter FROM public.trees WHERE street_tree = true ORDER BY id LIMIT {self.batch_size} OFFSET {self.batch_size*self.batch_num}", self.engine.connect())
```

Comment:
Auch das hab ich im Code nicht perfekt durchgezogen. Allerdings ist es "bad practice" hier mit String Format zu arbeiten in SQL Queries. Gibt dann die gefahr von SQL-Injections.  In dem Fall egal, aber wenn z.B. Batch Size von nem User käme, könnte er hier SQL Code in die Variabe geben. 

Besser: mit den Funktionen der read_sql arbeiten oder von connections. Das sieht zwar ähnlich aus, behandelt aber die SQL injections. 

Ich hab das z.B. in script_store_solaranywhere_forecast_in_db copy so gelöst:
                ```
prepared_query = "DELETE FROM private.weather_tile_forecast WHERE tile_id = %(tile_id)s AND date >= %(start_date)s AND date < %(end_date)s"
                with engine.connect() as connection:
                    connection.execute(prepared_query, tile_id=loc[0], start_date=start, end_date=end)

```

Mit der Pandas Funktion geht das so:
                ```
current_weather = pd.read_sql("SELECT DISTINCT ON (date) date, temp_max_c, rainfall_mm FROM private.weather_tile_forecast WHERE date = %s ORDER BY date, created_at DESC", 
                                            engine, params=(forecast_date, ))
```

Sieht ähnlich aus, ist aber sauberer als mit Stringformat zu arbeiten.

Reply:
Ich hatte ein bisschen Probleme mit WHERE IN tuple(). Irgendwie mag read_sql es nicht, wenn man str(tuple(abc)) als parameter mitgibt. Habe jetzt jeden Baum einzeln geladen und dann die Dataframes concatenated. Kostet etwas performance, vielleicht geht das auch noch besser